feat(events): add audit trail feature

sahil143 opened the pull request:
<strong>Note: If there are pending changes to the PR, prefix the PR title with "WIP" and add the label "DO NOT MERGE"</strong>

### Mandatory
 - [ ] What does this PR do? 
 Implement Audit-Trail feature

- [ ] What issue/task does this PR references?
https://openshift.io/openshiftio/openshiftio/plan/detail/1915

- [ ] Are the tests Included?
Added tests
___
### Optional
- [ ] Is the documentation Included?

- [ ] Are the Release Notes included?
<!-- For inclusion in marketing announcement - N/A for bugs. -->
<!-- A brief two line documentation of the functionality added/improved -->

- [ ] @mention(s) to expected reviewer(s) included

**Note:**
Document for Audit-Trail messages: https://docs.google.com/spreadsheets/d/1fi2ZbAECorWkM3Krug7HhyXsUdF4wa76RWTPmPMhDkU/edit?ts=5b080e62#gid=423213162

jarifibrahim commented on src/app/components_ngrx/work-item-event-wrapper/work-item-event-wrapper.module.ts:
```diff
@@ -0,0 +1,48 @@
+
+let providers = [];
+
+if (process.env.ENV == 'inmemory') {
```
Why `inmemory` mode?
We no longer use `inmemory` mode.

divyanshiGupta commented on src/app/components_ngrx/work-item-event-wrapper/work-item-event-wrapper.component.html:
```diff
@@ -0,0 +1,32 @@
+        *ngFor="let event of events | async"
+        [event]="event"
+        [type]="event.type"
+        >
```
You can remove this extra line and add `>` in line 28

divyanshiGupta commented on src/app/components_ngrx/work-item-event-wrapper/work-item-event-wrapper.component.ts:
```diff
@@ -0,0 +1,27 @@
+
+  private events = this.eventsQuery.getEventsWithModifier();
+
+    constructor( private store: Store < AppState >,
```
Indent the constructor properly and remove extra spaces in the parameters of the constructor, it should be
`private store: Store<AppState>`

divyanshiGupta commented on src/app/components_ngrx/work-item-event/work-item-event.component.html:
```diff
@@ -0,0 +1,65 @@
+        placement="right"
+        container="body"
+        *ngIf="(event.modifier | async).avatar"
+        [src]="(event.modifier | async).avatar | almAvatarSize:30" />
```
remove extra space before closing the tag

divyanshiGupta commented on src/app/components_ngrx/work-item-event/work-item-event.component.less:
```diff
@@ -0,0 +1,20 @@
+@import (reference) '../../../assets/stylesheets/_base';
+@import './../work-item-comment/work-item-comment.component.less';
+
+
```
remove extra line

divyanshiGupta commented on src/app/components_ngrx/work-item-event/work-item-event.component.ts:
```diff
@@ -0,0 +1,112 @@
+
+
+
+}
```
remove extra lines

maxandersen commented:
I tried to spot it in code but couldn't - is this using the existing audit logs backend already does for all table updates implemented by @xcoulon a while back or is this introducing something new ? 

And how does https://docs.google.com/spreadsheets/d/1fi2ZbAECorWkM3Krug7HhyXsUdF4wa76RWTPmPMhDkU/edit?ts=5b080e62#gid=423213162 relate ?

sahil143 commented:
>I tried to spot it in code but couldn't - is this using the existing audit logs backend already does for all table  updates implemented by @xcoulon a while back or is this introducing something new ?

@maxandersen Dont know if @xcoulon implementation is being used by backend or not. But, This ([Events](https://github.com/fabric8-services/fabric8-wit/pull/1959)) was the Pr that introduced the events api.
@baijum ^

The spreadsheet contains all the text formats used to show the audit on the particular field of the work item on the UI.

Raunak1203 commented:
@sahil143 if i update iteration/Area from parent to child, logs shown are not correct, 
- Changed the area from 'child area' to 'new_test_space' 
but logs shows "_Raunak  moved this work item from area 'new_test_space' to area 'child area_" 

![peek 2018-05-28 22-19](https://user-images.githubusercontent.com/34271052/40623989-976ac4a2-62c7-11e8-9de7-52dfc5893fe5.gif)

sahil143 commented:
> Shouldn't the timeline changes be collapsed bydefault(right angle) ?

we are going to merge comments and audit trail together, So it will be open by default, I have added the empty text for this one when there is no event.

> I can't see logs for Added/removed links from a workitem.

We are not capturing logs for links 

> In audit logs, Labels are clickable, If i click a label, results are filtered based on label condition.

fixed this

> It make sense if we add a label but if we remove it, it doesn't make sense.

can you explain this? I dont understand

> Changed the area from 'child area' to 'new_test_space'
but logs shows "Raunak moved this work item from area 'new_test_space' to area 'child area"

I know I already fixed this. Text should be "Raunak moved this work item to area 'new_test_space' from area 'child area"

> And also in the detail page clicking label is appending the URL but no change in the detail page as expected.

This should not happen anymore

@Raunak1203 ^^
And Please review again
